command/server_util.go

```diff
@@ -14,63 +13,78 @@ var (
-	if existBarrierSealConfig != nil && existBarrierSealConfig.Type != vaultseal.HSMAutoDeprecated &&
-		(existBarrierSealConfig.Type != seal.BarrierType() ||
-			config.Seal != nil && config.Seal.Disabled) {
-		// If the existing seal is not Shamir, we're going to Shamir, which
-		// means we require them setting "disabled" to true in their
-		// configuration as a sanity check.
-		if (config.Seal == nil || !config.Seal.Disabled) && existBarrierSealConfig.Type != vaultseal.Shamir {
-			return errors.New(`Seal migration requires specifying "disabled" as "true" in the "seal" block of Vault's configuration file"`)
-		}
 
-		// Conversely, if they are going from Shamir to auto, we want to
-		// ensure disabled is *not* set
-		if existBarrierSealConfig.Type == vaultseal.Shamir && config.Seal != nil && config.Seal.Disabled {
-			coreConfig.Logger.Warn(`when not migrating, Vault's config should not specify "disabled" as "true" in the "seal" block of Vault's configuration file`)
+	// If we don't have an existing config or if it's the deprecated auto seal
+	// which needs an upgrade, skip out
+	if existBarrierSealConfig == nil || existBarrierSealConfig.Type == vaultseal.HSMAutoDeprecated {
+		return nil
+	}
+
+	if unwrapSeal == nil {
+		// We have the same barrier type and the unwrap seal is nil so we're not
+		// migrating from same to same, IOW we assume it's not a migration
+		if existBarrierSealConfig.Type == barrierSeal.BarrierType() {
 			return nil
 		}
 
-		if existBarrierSealConfig.Type != vaultseal.Shamir && existRecoverySealConfig == nil {
-			return errors.New(`Recovery seal configuration not found for existing seal`)
+		// If we're not coming from Shamir, and the existing type doesn't match
+		// the barrier type, we need both the migration seal and the new seal
+		if existBarrierSealConfig.Type != vaultseal.Shamir && barrierSeal.BarrierType() != vaultseal.Shamir {
+			return errors.New(`Trying to migrate from auto-seal to auto-seal but no "disabled" seal stanza found`)
+		}
+	} else {
+		if unwrapSeal.BarrierType() == vaultseal.Shamir {
+			return errors.New("Shamir seals cannot be set disabled (they should simply not be set)")
 		}
+	}
 
-		switch existBarrierSealConfig.Type {
-		case vaultseal.Shamir:
-			// The value reflected in config is what we're going to
-			existSeal = vault.NewDefaultSeal()
-			existSeal.SetCore(core)
-			newSeal = seal
-			newBarrierSealConfig := &vault.SealConfig{
-				Type:            newSeal.BarrierType(),
-				SecretShares:    1,
-				SecretThreshold: 1,
-				StoredShares:    1,
-			}
-			newSeal.SetCachedBarrierConfig(newBarrierSealConfig)
-			newSeal.SetCachedRecoveryConfig(existBarrierSealConfig)
+	var existSeal vault.Seal
+	var newSeal vault.Seal
+
+	if existBarrierSealConfig.Type == barrierSeal.BarrierType() {
+		// In this case our migration seal is set so we are using it
+		// (potentially) for unwrapping. Set it on core for that purpose then
+		// exit.
+		core.SetSealsForMigration(nil, nil, unwrapSeal)
```

Comment:
We are still correctly setting core.seal to the operating seal sometime after this, right?

Reply:
c.seal only gets changed if the first two arguments *aren't* nil so yes.